### PR TITLE
Cleanup UState API usage

### PR DESCRIPTION
**Pull request by SkySkimmer**
- deprecate those evar_universe_context aliases in Evd
- rename UState normalize -> minimize, Evd nf_constraints -> minimize_universes
- deprecate those functions which mix minimization and evar-delta normalization (like nf_evars_and_universes) letting the callers replace them by 2 calls (sometimes just 1 when they don't use the substitution)
- stop using Universes.nf_evars_and_universes_opt_subst. 

The first 2 are just renames.
There is a possible performance impact with the later 2 as Universes.nf_evars_and_universes_opt_subst does some tricky stuff updating a ref of the substitution (which is not done by the replacement using to_constr), we will see what happens.

### CODE REVIEW

File: proofs/proof_global.ml [hunk 1]
```diff
@@ -336,10 +336,10 @@ let close_proof ~keep_body_ucst_separate ?feedback_id ~now
...
      have existential variables in the initial types of goals, we need to
      normalise them for the kernel. *)
-  let subst_evar k =
-    Proof.in_proof proof (fun m -> Evd.existential_opt_value m k) in
-  let nf = Universes.nf_evars_and_universes_opt_subst subst_evar
-    (Evd.evar_universe_context_subst universes) in
+  let nf = Proof.in_proof proof (fun sigma ->
+      let sigma = Evd.set_universe_context sigma universes in
+      Evarutil.nf_evars_universes sigma)
+  in
```
Comment (SkySkimmer) [inline proofs/proof_global.ml]: I'm not sure if this is cleaner with nf_evars_and_universes_opt_subst or using Evd.set_universe_context, but keeping around a separate implementation of EConstr.to_constr is a bit weird.

**Comment by SkySkimmer** [comment]
https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/361/console

```


┌──────────────────────────┬─────────────────────────┬─────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                          │      user time [s]      │             CPU cycles              │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                          │                         │                                     │                                       │                         │                 │
│             package_name │     NEW     OLD PDIFF   │           NEW           OLD PDIFF   │            NEW            OLD PDIFF   │     NEW     OLD PDIFF   │ NEW OLD PDIFF   │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│         coq-fiat-parsers │  657.80  662.80 -0.75 % │ 1835581096081 1847757526239 -0.66 % │  2947202144658  2958655317926 -0.39 % │ 3345740 3351172 -0.16 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-ssreflect │   40.37   40.66 -0.71 % │  111372803201  111459572245 -0.08 % │   137209880564   137373091007 -0.12 % │  478432  478816 -0.08 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-formal-topology │   34.17   34.41 -0.70 % │   94802076465   95036819789 -0.25 % │   124946799551   125121070460 -0.14 % │  483664  483688 -0.00 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-solvable │  189.97  190.98 -0.53 % │  527003959159  528140844105 -0.22 % │   760333705861   761410782095 -0.14 % │  709516  710700 -0.17 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-bignums │   75.04   75.42 -0.50 % │  209118579138  208911539899 +0.10 % │   282687560363   283425845883 -0.26 % │  524164  526600 -0.46 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│           coq-coquelicot │   75.34   75.64 -0.40 % │  207735278327  207941028969 -0.10 % │   259453156637   259921167807 -0.18 % │  656756  656456 +0.05 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│             coq-compcert │  812.55  815.18 -0.32 % │ 2266223276257 2275209348358 -0.39 % │  3419699073236  3425595465920 -0.17 % │ 1363120 1369280 -0.45 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-flocq │   60.02   60.14 -0.20 % │  164386113648  164014033080 +0.23 % │   211348086374   211574370309 -0.11 % │  662372  660732 +0.25 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-geocoq │ 3044.76 3046.81 -0.07 % │ 8466999453289 8476130642575 -0.11 % │ 13958344376973 13968577444259 -0.07 % │ 1241844 1241924 -0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│     coq-mathcomp-algebra │  169.06  168.97 +0.05 % │  468198820419  467173547957 +0.22 % │   656527464736   655910523770 +0.09 % │  585996  588816 -0.48 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-odd_order │ 1344.57 1343.52 +0.08 % │ 3734755601059 3733024816358 +0.05 % │  6590808119754  6591073612626 -0.00 % │ 1077992  988820 +9.02 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-hott │  309.77  309.50 +0.09 % │  844307888767  842306606971 +0.24 % │  1376326345957  1376010673734 +0.02 % │  496768  496728 +0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-corn │ 1481.08 1479.57 +0.10 % │ 4111573211365 4107078674521 +0.11 % │  6396561567651  6400086726522 -0.06 % │  889132  835940 +6.36 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-mathcomp-field │  441.42  440.91 +0.12 % │ 1225340954157 1223618598412 +0.14 % │  2023174357460  2025017321425 -0.09 % │  712632  713884 -0.18 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-character │  262.96  262.53 +0.16 % │  730759466785  729867265682 +0.12 % │  1074102774795  1074605855691 -0.05 % │  981352  983064 -0.17 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│ coq-mathcomp-real_closed │  166.34  165.76 +0.35 % │  460888969197  459876977196 +0.22 % │   699125489000   699216601176 -0.01 % │  716900  714644 +0.32 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-color │  571.40  566.16 +0.93 % │ 1590209710142 1581981893722 +0.52 % │  1988879869986  1990250809088 -0.07 % │ 1439572 1438672 +0.06 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-fingroup │   60.78   60.18 +1.00 % │  167100947078  166458661002 +0.39 % │   231040092841   231369518167 -0.14 % │  525988  529076 -0.58 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│            coq-fiat-core │  102.91  101.58 +1.31 % │  290629779334  289761891511 +0.30 % │   377721737973   378268710879 -0.14 % │  501612  501228 +0.08 % │   0   0  +nan % │
└──────────────────────────┴─────────────────────────┴─────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘


INFO: the following OPAM-packages:
- coq-fiat-crypto
- coq-sf-lf
- coq-sf-plf
- coq-sf-vfa
- coq-unimath
- coq-vst
cannot be installed (exit 1)

```

**Comment by maximedenes** [comment]
@SkySkimmer if this is still on your schedule for 8.8+beta1, can you rebase? @ppedrot @ejgallego can you review?

If not, can you update the milestone? Thanks!

**Comment by SkySkimmer** [comment]
Rebased.

**Review by ejgallego — APPROVED** [review]
I'm not an expert but this looks like a great and much needed cleanup. cc #6414

**Review by ppedrot — APPROVED** [review]
Looks to me like a good cleanup over the current universe API state.

**Comment by SkySkimmer** [comment]
It seems this breaks fiat-crypto:
- https://travis-ci.org/SkySkimmer/coq/jobs/349796288 out of memory
- https://circleci.com/gh/coq/coq/16348 "Command terminated by signal 9"

I'm going to delay the last 2 commits to a 8.9 PR. The first 2 are function renamings only with deprecated aliases available so should not be breaking in any way.

**Comment by maximedenes** [comment]
Seems ok: https://travis-ci.org/SkySkimmer/coq/builds/350289983

**Pull request merged**